lucene/core/src/java/org/apache/lucene/util/IntroSelector.java

```diff
@@ -33,26 +32,87 @@ public final void select(int from, int to, int k) {
     quickSelect(from, to, k, maxDepth);
   }
 
-  // heap sort
-  // TODO: use median of median instead to have linear worst-case rather than
-  // n*log(n)
-  void slowSelect(int from, int to, int k) {
-    new Sorter() {
+  int slowSelect(int from, int to, int k) {
+    return medianOfMediansSelect(from, to-1, k);
+  }
+
+  int medianOfMediansSelect(int from, int to, int k) {
+    int pivotIndex;
+    do {
+      if (from == to) {
```

Comment:
I believe this condition can never happen when `medianOfMediansSelect` is called from `quickSelect` since it first checks whether `to-from==1` before delegating to `medianOfMediansSelect`? It is probably a good idea to be defensive, but maybe leave a comment about it?